lib/DB/Schema/ResultSet/User.pm

```diff
@@ -105,10 +105,13 @@ sub addGlobalUser {
 	my ( $self, $user_params, $as_result_set ) = @_;
 	DB::Exception::ParametersNeeded->throw( message => "The parameters must include username" )
 		unless defined( $user_params->{username} );
+	my $params = clone($user_params);
+	# remove the user_id if defined and equal to zero.
+	if ( defined ($params->{user_id}) && $params->{user_id} == 0) {
+		delete $params->{user_id};
+	}
```

Comment:
The `user_id` field is autogenerated.  On the client side when we make a new user, I assume the `user_id` to be 0.  However, on the DB side, if it is zero, it's a new user, so delete it and let the DB generate the `user_id`. 

Alternatively (and as I write this it makes more sense), to not assign a `user_id` on the client side.  I think I had reason for not doing that, but I can't remember. 

If we keep with the new user has `user_id` of 0, I'll add a comment to the code.

Reply:
Right.  I am confusing the `user_id` with the `username`.  With webwork2 there was no `username`, and the `user_id` played that role.

I agree that the `user_id` probably should not be allowed to be set from the client, and should always be auto generated.  

There may be some issues with these database ids when it comes to course archival and restoration though.  If a course is archived, do we save these database ids, and then also restore them?  That could be problematic if during the time that the course is archived another user is created that has the same id.   If the course is archived and restored and the ids change, then the browser url for pages the user accesses will change, so that could be a problem if these ids are not saved and restored.  We may need to rethink using route params in the way that you have the vue router set up.

The route param issue has been on my mind for a while.  This may also affect things like LTI authentication.

Reply:
About the archiving, I don't think we can keep the database ids, because there may be clashes with existing one.  I think if we import a course (for example), we add the users as new users and let the DB regenerate new ids. 

Say more about the route param issue?

Reply:
What you are talking about is exactly the route param issue.  If a course is archived and then restored with this plan, then the urls for a user will all possibly have changed due to the course_id and user_id having changed.  For example, before archival the route may have been `/webwork3/courses/4/instructor`, and after restoring at some later time it will now be perhaps `/webwork3/courses/8/instructor` assuming several courses were added in the mean time.  This has implications for linking into webwork externally that we are going to need to figure out.

Reply:
I see what you mean.  Alternative to the course_id we could use the course name in the route.  That would be more robust.

Reply:
Perhaps, but that may not work so well either because with the new database scheme course names can have characters that might be problematic.